File: backend/models/UserProfile.js
const prisma = require('../lib/prisma');

class UserProfile {
  constructor(data) {
    this.id = data.id;
    this.height = data.height;
    this.weight = data.weight;
    this.prePregnancyWeight = data.prePregnancyWeight;
    this.age = data.age;
    this.gender = data.gender || 'female';
    this.locality = data.locality;
    this.timezone = data.timezone;
    this.medicalHistory = data.medicalHistory;
    this.allergies = data.allergies;
    this.medications = data.medications;
    this.lifestyle = data.lifestyle;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
  }

  async save() {
    const data = {
      height: this.height,
      weight: this.weight,
      prePregnancyWeight: this.prePregnancyWeight,
      age: this.age,
      gender: this.gender,
      locality: this.locality,
      timezone: this.timezone,
      medicalHistory: this.medicalHistory,
      allergies: this.allergies,
      medications: this.medications,
      lifestyle: this.lifestyle,
    };

    if (this.id) {
      // Update existing profile
      const updated = await prisma.userProfile.update({
        where: { id: this.id },
        data: data,
      });
      return new UserProfile(updated);
    } else {
      // Create new profile
      const created = await prisma.userProfile.create({
        data: data,
      });
      return new UserProfile(created);
    }
  }

  static async get() {
    const profile = await prisma.userProfile.findFirst();
    
    if (profile) {
      return new UserProfile(profile);
    }
    
    return null;
  }

  static async update(updates) {
    const existingProfile = await UserProfile.get();
    if (!existingProfile) {
      throw new Error('No user profile found');
    }

    const updated = await prisma.userProfile.update({
      where: { id: existingProfile.id },
      data: updates,
    });

    return new UserProfile(updated);
  }

  static async delete() {
    const existingProfile = await UserProfile.get();
    if (!existingProfile) {
      throw new Error('No user profile found');
    }

    await prisma.userProfile.delete({
      where: { id: existingProfile.id },
    });

    return true;
  }

  // Helper methods
  getBMI() {
    if (this.height && this.weight) {
      const heightInMeters = this.height / 100;
      return (this.weight / (heightInMeters * heightInMeters)).toFixed(1);
    }
    return null;
  }

  getWeightGain() {
    if (this.prePregnancyWeight && this.weight) {
      return (this.weight - this.prePregnancyWeight).toFixed(1);
    }
    return null;
  }

  getFormattedProfile() {
    const profile = [];
    
    if (this.age) profile.push(`Age: ${this.age} years`);
    if (this.height) profile.push(`Height: ${this.height} cm`);
    if (this.weight) profile.push(`Current weight: ${this.weight} kg`);
    if (this.prePregnancyWeight) profile.push(`Pre-pregnancy weight: ${this.prePregnancyWeight} kg`);
    if (this.getBMI()) profile.push(`BMI: ${this.getBMI()}`);
    if (this.getWeightGain()) profile.push(`Weight gain: ${this.getWeightGain()} kg`);
    if (this.locality) profile.push(`Location: ${this.locality}`);
    if (this.timezone) profile.push(`Timezone: ${this.timezone}`);
    
    return profile.join(', ');
  }

  getMedicalContext() {
    const context = [];
    
    if (this.medicalHistory && Array.isArray(this.medicalHistory) && this.medicalHistory.length > 0) {
      context.push(`Medical history: ${this.medicalHistory.join(', ')}`);
    }
    if (this.allergies && Array.isArray(this.allergies) && this.allergies.length > 0) {
      context.push(`Allergies: ${this.allergies.join(', ')}`);
    }
    if (this.medications && Array.isArray(this.medications) && this.medications.length > 0) {
      context.push(`Current medications: ${this.medications.join(', ')}`);
    }
    if (this.lifestyle && typeof this.lifestyle === 'object') {
      const lifestyleItems = [];
      if (this.lifestyle.diet) lifestyleItems.push(`Diet: ${this.lifestyle.diet}`);
      if (this.lifestyle.exercise) lifestyleItems.push(`Exercise: ${this.lifestyle.exercise}`);
      if (this.lifestyle.smoking) lifestyleItems.push(`Smoking: ${this.lifestyle.smoking}`);
      if (this.lifestyle.alcohol) lifestyleItems.push(`Alcohol: ${this.lifestyle.alcohol}`);
      if (lifestyleItems.length > 0) {
        context.push(`Lifestyle: ${lifestyleItems.join(', ')}`);
      }
    }
    
    return context.join('; ');
  }

  toJSON() {
    return {
      id: this.id,
      height: this.height,
      weight: this.weight,
      prePregnancyWeight: this.prePregnancyWeight,
      age: this.age,
      gender: this.gender,
      locality: this.locality,
      timezone: this.timezone,
      medicalHistory: this.medicalHistory,
      allergies: this.allergies,
      medications: this.medications,
      lifestyle: this.lifestyle,
      bmi: this.getBMI(),
      weightGain: this.getWeightGain(),
      createdAt: this.createdAt,
      updatedAt: this.updatedAt
    };
  }
}

module.exports = UserProfile;

File: luma/lib/models/user_profile.dart
import '../services/device_timezone_service.dart';

class UserProfile {
  final String id;
  final double? height; // in cm
  final double? weight; // in kg
  final double? prePregnancyWeight; // in kg
  final int? age;
  final String gender;
  final String? locality;
  final String? timezone;
  final List<String>? medicalHistory;
  final List<String>? allergies;
  final List<String>? medications;
  final Lifestyle? lifestyle;
  final DateTime createdAt;
  final DateTime updatedAt;

  UserProfile({
    required this.id,
    this.height,
    this.weight,
    this.prePregnancyWeight,
    this.age,
    this.gender = 'female',
    this.locality,
    this.timezone,
    this.medicalHistory,
    this.allergies,
    this.medications,
    this.lifestyle,
    required this.createdAt,
    required this.updatedAt,
  });

  // Helper methods
  double? get bmi {
    if (height != null && weight != null) {
      final heightInMeters = height! / 100;
      return weight! / (heightInMeters * heightInMeters);
    }
    return null;
  }

  double? get weightGain {
    if (prePregnancyWeight != null && weight != null) {
      return weight! - prePregnancyWeight!;
    }
    return null;
  }

  String get formattedProfile {
    final profile = <String>[];
    
    if (age != null) profile.add('Age: $age years');
    if (height != null) profile.add('Height: ${height!.toStringAsFixed(0)} cm');
    if (weight != null) profile.add('Current weight: ${weight!.toStringAsFixed(1)} kg');
    if (prePregnancyWeight != null) profile.add('Pre-pregnancy weight: ${prePregnancyWeight!.toStringAsFixed(1)} kg');
    if (bmi != null) profile.add('BMI: ${bmi!.toStringAsFixed(1)}');
    if (weightGain != null) profile.add('Weight gain: ${weightGain!.toStringAsFixed(1)} kg');
    if (locality != null) profile.add('Location: $locality');
    if (timezone != null) profile.add('Timezone: $timezone');
    
    return profile.join(', ');
  }

  String get medicalContext {
    final context = <String>[];
    
    if (medicalHistory != null && medicalHistory!.isNotEmpty) {
      context.add('Medical history: ${medicalHistory!.join(', ')}');
    }
    if (allergies != null && allergies!.isNotEmpty) {
      context.add('Allergies: ${allergies!.join(', ')}');
    }
    if (medications != null && medications!.isNotEmpty) {
      context.add('Current medications: ${medications!.join(', ')}');
    }
    if (lifestyle != null) {
      final lifestyleItems = <String>[];
      if (lifestyle!.diet != null) lifestyleItems.add('Diet: ${lifestyle!.diet}');
      if (lifestyle!.exercise != null) lifestyleItems.add('Exercise: ${lifestyle!.exercise}');
      if (lifestyle!.smoking != null) lifestyleItems.add('Smoking: ${lifestyle!.smoking}');
      if (lifestyle!.alcohol != null) lifestyleItems.add('Alcohol: ${lifestyle!.alcohol}');
      if (lifestyleItems.isNotEmpty) {
        context.add('Lifestyle: ${lifestyleItems.join(', ')}');
      }
    }
    
    return context.join('; ');
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'height': height,
      'weight': weight,
      'prePregnancyWeight': prePregnancyWeight,
      'age': age,
      'gender': gender,
      'locality': locality,
      'timezone': timezone,
      'medicalHistory': medicalHistory,
      'allergies': allergies,
      'medications': medications,
      'lifestyle': lifestyle?.toJson(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  factory UserProfile.fromJson(Map<String, dynamic> json) {
    return UserProfile(
      id: json['id'] ?? '',
      height: json['height']?.toDouble(),
      weight: json['weight']?.toDouble(),
      prePregnancyWeight: json['prePregnancyWeight']?.toDouble(),
      age: json['age'],
      gender: json['gender'] ?? 'female',
      locality: json['locality'],
      timezone: json['timezone'],
      medicalHistory: json['medicalHistory'] != null 
          ? List<String>.from(json['medicalHistory']) 
          : null,
      allergies: json['allergies'] != null 
          ? List<String>.from(json['allergies']) 
          : null,
      medications: json['medications'] != null 
          ? List<String>.from(json['medications']) 
          : null,
      lifestyle: json['lifestyle'] != null 
          ? Lifestyle.fromJson(json['lifestyle']) 
          : null,
      createdAt: DeviceTimezoneService.toDeviceTimezone(DateTime.parse(json['createdAt'])),
      updatedAt: DeviceTimezoneService.toDeviceTimezone(DateTime.parse(json['updatedAt'])),
    );
  }

  UserProfile copyWith({
    String? id,
    double? height,
    double? weight,
    double? prePregnancyWeight,
    int? age,
    String? gender,
    String? locality,
    String? timezone,
    List<String>? medicalHistory,
    List<String>? allergies,
    List<String>? medications,
    Lifestyle? lifestyle,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return UserProfile(
      id: id ?? this.id,
      height: height ?? this.height,
      weight: weight ?? this.weight,
      prePregnancyWeight: prePregnancyWeight ?? this.prePregnancyWeight,
      age: age ?? this.age,
      gender: gender ?? this.gender,
      locality: locality ?? this.locality,
      timezone: timezone ?? this.timezone,
      medicalHistory: medicalHistory ?? this.medicalHistory,
      allergies: allergies ?? this.allergies,
      medications: medications ?? this.medications,
      lifestyle: lifestyle ?? this.lifestyle,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

class Lifestyle {
  final String? diet;
  final String? exercise;
  final String? smoking;
  final String? alcohol;

  Lifestyle({
    this.diet,
    this.exercise,
    this.smoking,
    this.alcohol,
  });

  Map<String, dynamic> toJson() {
    return {
      'diet': diet,
      'exercise': exercise,
      'smoking': smoking,
      'alcohol': alcohol,
    };
  }

  factory Lifestyle.fromJson(Map<String, dynamic> json) {
    return Lifestyle(
      diet: json['diet'],
      exercise: json['exercise'],
      smoking: json['smoking'],
      alcohol: json['alcohol'],
    );
  }

  Lifestyle copyWith({
    String? diet,
    String? exercise,
    String? smoking,
    String? alcohol,
  }) {
    return Lifestyle(
      diet: diet ?? this.diet,
      exercise: exercise ?? this.exercise,
      smoking: smoking ?? this.smoking,
      alcohol: alcohol ?? this.alcohol,
    );
  }
}
